### PR TITLE
responsive compiler prototype: build system changes

### PR DESCRIPTION
**Pull request by mppf**
This PR contains build changes for the prototype compiler library effort.
The Makefile targets are not run yet and refer to files that will be
added in other PRs (#17586, #17587, #17588, #17589).

This PR is split out from PR #17583, which is a start at the compiler
revamp effort described in the [1.24 release notes ongoing
efforts](https://chapel-lang.org/releaseNotes/1.24/04-ongoing.pdf).

 * The new effort uses `cmake`
 * The new code includes docs comments with doxygen+breathe+sphinx. As
   such, it adds `breathe` as a Python requirement for `chpldoc`. (These
   requirements are normally automatically installed). For now, `make
   docs` will continue to work if `doxygen` is not installed and has not
   been run on the compiler sources. To support that, there are actually
   two `index.rst` files for `compiler-internals` in the `doc/templates`
   directory. The Makefiles will use one or the other based upon whether
   or not the doxygen materials are found. In the future we might want to
   simplify the situation by requiring `doxygen` in order for `make docs`
   to succeed.
  * The new effort uses a different directory structure from the old
    effort. For now it is stored in `compiler/next` but the expectation
    is that eventually we will move all of the compiler code into this
    directory structure. The change in directory structure is described a
    bit in
    [compiler/next/README](https://github.com/chapel-lang/chapel/blob/53c5190306bbe91fb3db221ff28d510f98b51b3a/compiler/next/README).
 * There are now C++ tests of portions of the compiler but since this PR
   doesn't add any C++ sources they can't run yet. When they are present
   though they can be built and run with `make test-libchplcomp`.
 * I was having trouble debugging the tests linked with a dynamic
   library. So, the tests to using a CMake object library for these
   tests. That just means that the test source code is compiled and
   linked with the `.o` files in the library (rather than a library `.a`
   or `.so). The object library is also useful because it allows one to
   create both a `.a` and a `.so` for the same library. Anyway, the
   library itself builds with the default of shared/static for the build
   library (and the default depends on the BUILD_SHARED_LIBS setting).
 * While there, updated `make test` to run only the tests in examples/
   because what it had would run all of the tests (which takes something
   like a day). Note that `make test` is only available in source
   checkouts and not included in releases.

 - [x] full local testing
 
 Reviewed by @lydia-duncan @gbtitus @dlongnecke-cray - thanks!

### CODE REVIEW

**Comment by lydia-duncan** [comment]
>     * So, the tests to using an object library (always) for these tests.

Looks like there's a missing word here?

File: Makefile.devel [hunk 1]
```diff
@@ -58,8 +58,17 @@ always-build-chplspell-venv: FORCE
 	$(MAKE) chplspell-venv; \
 	fi
 
-test: FORCE
-	cd test && start_test
+test: spectests FORCE
+	cd examples && start_test
```
Comment (lydia-duncan) [inline Makefile.devel:62]: This looks like an unrelated change and I don't see it called out in the PR message

File: compiler/next/README [hunk 2]
```diff
@@ -0,0 +1,76 @@
+This directory contains a start towards work on the revamped compiler.
+This directory uses a different directory structure from the old compiler
+- see below for rationale.
+
+For now, this work is being done in this directory but the expectation is
+that eventually it swap places with the compiler/ directory (so any
```
Comment (lydia-duncan) [inline compiler/next/README:6]: "eventually it swap" -> "eventually it will swap"

File: compiler/next/README [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+old-compiler code still used is put into subdirectories according to the
+organization described below).
+
+
+As a preparatory step for the effort to significantly improve the
+compiler, thinking about the directory structure we use for sources
```
Comment (lydia-duncan) [inline compiler/next/README:12]: There's a missing word here but I'm not sure what it is

**Review by lydia-duncan — COMMENTED** [review]
I don't really feel confident in my Makefile-fu so it'd be good if maybe @gbtitus reviewed those parts.  I did call out one thing that struck me as off and took a look at the README

File: Makefile.devel (hunk 1 above)
Comment (mppf) [reply Makefile.devel:62]: I'll add something to the PR message along these lines:

While there, updated `make test` to run the tests in examples/ because what it had would run all of the tests (which takes something like a day). Note that `make test` is only available in source checkouts and not included in releases.

Comment (lydia-duncan) [reply Makefile.devel:62]: Awesome, thanks!

File: Makefile [hunk 4]
```diff
@@ -73,6 +73,10 @@ notcompiler: FORCE
 	@$(MAKE) runtime
 	@$(MAKE) modules
 
+libchplcomp: FORCE
+	@echo "Making the compiler library..."
+	@cd compiler/next && $(MAKE) -f Makefile.help libchplcomp
```
Comment (gbtitus) [inline Makefile:78]: You'll make life easier for yourself in the future if you:

- Just do `@cd compiler && $(MAKE) $@` here. `$@` is automatically set and names the rule's target.
- Add a `libchplcomp` rule to `compiler/Makefile` that does `cd $LIBCHPLDIR && $(MAKE) -f Makefile.help $@`. This is the rule invoked by the above.
- Add a definition `LIBCHPLDIR=next` to `compiler/make/Makefile.compiler.head`.

This will let you change the location of the libchpl directory just by changing that `LIBCHPLDIR` variable, instead of having to look all through the makefiles and find all the instances of 'next'.

File: Makefile.devel [hunk 5]
```diff
@@ -58,8 +58,17 @@ always-build-chplspell-venv: FORCE
...
+
+libchplcomp-docs: FORCE
+	@cd compiler/next && $(MAKE) -f Makefile.help libchplcomp-docs
```
Comment (gbtitus) [inline Makefile.devel:65]: You could do the same thing for these three as I described for `Makefile:libchplcomp`.

File: compiler/next/Makefile.help [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+COMPILER_ROOT = ..
+COMPILER_SUBDIR = next
+
+#
+# standard header
+#
+include $(COMPILER_ROOT)/make/Makefile.compiler.head
+
+LIB_CMAKE_ARG := -DCMAKE_BUILD_TYPE=Release
+LIBCOMPILER_BUILD_DIR := $(COMPILER_BUILD)/next/Release
```
Comment (gbtitus) [inline compiler/next/Makefile.help:49]: Could use `$(LIBCHPLDIR)` instead of 'next' here, if it were defined in `compiler/make/Makefile.compiler.head`.

**Comment by gbtitus** [comment]
I didn't spot anything that seemed broken or wrong in the GMakefiles, but described an idea in line-comments that would ease future transitions. I didn't look at the CMakefiles, because I lack all fu with that tool.
👍

**Review by gbtitus — APPROVED** [review]

File: Makefile (hunk 4 above)
Comment (mppf) [reply Makefile:78]: I think these are interesting ideas but I think it's better not to do it yet in this case. One reason is that I'm hoping in the future we swap`compiler/` and `compiler/next` - i.e. we move the existing compiler code into the new directory structure. In that case, the change will be harder than just renaming `next` and it'll be easier if these rules are more straightforward.

**Review by dlongnecke-cray — APPROVED** [review]
The CMake ~~changes~~ stuff looks good to me!
